### PR TITLE
Add IoT Hub Tests

### PR DESCRIPTION
**Pull request by danewalton**

### CODE REVIEW

**Comment by hihigupt** [comment]
Number right now looks to be 43% can we increase the number of line coverage?
https://github.com/Azure/azure-iot-middleware-freertos/pull/83/checks?check_run_id=2336524259#step:5:204

**Comment by danewalton** [comment]
> Number right now looks to be 43% can we increase the number of line coverage?
> https://github.com/Azure/azure-iot-middleware-freertos/pull/83/checks?check_run_id=2336524259#step:5:204

yup will get right on that

File: tests/azure_iot_hub_client_ut.c [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+#include <cmocka.h>
+
+#include "azure_iot_mqtt.h"
+#include "azure_iot_hub_client.h"
+
+#define testCLOUD_CALLBACK_ID     1
+#define testMETHOD_CALLBACK_ID    2
+#define testTWIN_CALLBACK_ID      3
+#define testEMPTY_JSON            "{}"
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:19]: Can we remove this as this is more part of receive path.

File: tests/azure_iot_hub_client_ut.c [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+typedef struct ReceviceTestData
+{
+    const uint8_t * pucTopic;
+    uint32_t ulTopicLength;
+    const uint8_t * pucPayload;
+    uint32_t ulPayloadLength;
+    uint32_t ulCallbackFunctionId;
+} ReceviceTestData_t;
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:28]: nit: can we bring this from sdk_spike branch later?

File: tests/azure_iot_hub_client_ut.c [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+/*-----------------------------------------------------------*/
+TickType_t xTaskGetTickCount( void );
+uint32_t ulGetAllTests();
+
+TickType_t xTaskGetTickCount( void )
+{
+    return 1;
+}
+
+static uint64_t prvGetUnixTime( void )
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:56]: nit: ------

File: tests/azure_iot_hub_client_ut.c [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+static void testAzureIoTHubClient_Init_Failure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    AzureIoTHubClientOptions_t xHubClientOptions = { 0 };
+
+    ( void ) state;
+
+    /* Fail init when null hostname passed */
+    assert_int_not_equal( AzureIoTHubClient_Init( &xTestIoTHubClient,
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:87]: Can we add NULL handle test

File: tests/azure_iot_hub_client_ut.c [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+                                                  ucHostname, sizeof( ucHostname ) - 1,
+                                                  ucDeviceId, sizeof( ucDeviceId ) - 1,
+                                                  &xHubClientOptions,
+                                                  NULL, 0,
+                                                  prvGetUnixTime,
+                                                  &xTransportInterface ),
+                          eAzureIoTHubClientSuccess );
+
+    /* Fail init when null AzureIoTGetCurrentTimeFunc_t passed */
+    assert_int_not_equal( AzureIoTHubClient_Init( &xTestIoTHubClient,
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:119]: >assert_int_not_equal [](start = 4, length = 20)

lets check for equal for exact behaviour we our checking

File: tests/azure_iot_hub_client_ut.c [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+                      eAzureIoTHubClientSuccess );
+}
+
+static void testAzureIoTHubClient_Deinit_Success( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    AzureIoTHubClient_Deinit( &xTestIoTHubClient );
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:181]: Can we init and then do deinit?
Something like this:
{
    AzureIoTHubClient_t xTestIoTHubClient;

    ( void ) state;

 prvSetupTestIoTHubClient( &xTestIoTHubClient );

AzureIoTHubClient_Deinit( &xTestIoTHubClient );
}

File: tests/azure_iot_hub_client_ut.c [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+static void testAzureIoTHubClient_Connect_InvalidArgFailure( void ** state )
+{
+    ( void ) state;
+
+    bool xSessionPresent;
+
+    assert_int_not_equal( AzureIoTHubClient_Connect( NULL,
+                                                     false,
+                                                     &xSessionPresent,
+                                                     60 ),
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:193]: Can we pass xSessionPresent as null ?

File: tests/azure_iot_hub_client_ut.c [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+                          eAzureIoTHubClientSuccess );
+}
+
+static void testAzureIoTHubClient_Connect_MQTTConnectFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    bool xSessionPresent;
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:203]: >xSessionPresent [](start = 9, length = 15)

nit: start of function

File: tests/azure_iot_hub_client_ut.c (hunk 7 above)
Comment (hihigupt) [reply tests/azure_iot_hub_client_ut.c:193]: Also please add comment 

---
In reply to: [612734880](https://github.com/Azure/azure-iot-middleware-freertos/pull/83#discussion_r612734880) [](ancestors = 612734880)

File: tests/azure_iot_hub_client_ut.c (hunk 8 above)
Comment (hihigupt) [reply tests/azure_iot_hub_client_ut.c:203]: all the places

---
In reply to: [612735034](https://github.com/Azure/azure-iot-middleware-freertos/pull/83#discussion_r612735034) [](ancestors = 612735034)

File: tests/azure_iot_hub_client_ut.c [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+                          eAzureIoTHubClientSuccess );
+}
+
+static void testAzureIoTHubClient_SendTelemetry_BigTopicFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    AzureIoTMessageProperties_t properties =
+    {
+        ._internal.xProperties._internal.current_property_index = 0,
+        ._internal.xProperties._internal.properties_written     = sizeof( ucBuffer ) + 1
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:288]: >     [](start = 60, length = 4)

nit: remove extra space

File: tests/azure_iot_hub_client_ut.c (hunk 5 above)
Comment (danewalton) [reply tests/azure_iot_hub_client_ut.c:119]: yea I was going to say the same thing. done

File: tests/azure_iot_hub_client_ut.c (hunk 9 above)
Comment (danewalton) [reply tests/azure_iot_hub_client_ut.c:288]: this is using uncrustify so I'll just let it run its course

**Comment by danewalton** [comment]
@hihigupt the vast majority of the low code cov is the functions that aren't used yet.

File: tests/azure_iot_hub_client_ut.c [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+    return 1;
+}
+
+/*-----------------------------------------------------------*/
+
+static uint64_t prvGetUnixTime( void )
+{
+    return 0xFFFFFFFFFFFFFFFF;
+}
+
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:52]: nit:
/*--------------------------------*/
<blank line>
fun()
{
}
/*------------------------------*/
<blank line>

File: tests/azure_iot_hub_client_ut.c [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+static void testAzureIoTHubClient_Init_Success( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    AzureIoTHubClientOptions_t xHubClientOptions = { 0 };
+
+    ( void ) state;
+    will_return( AzureIoTMQTT_Init, eAzureIoTMQTTSuccess );
+
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:168]: nit:remove blank line and add line between 166 and 167

File: tests/azure_iot_hub_client_ut.c [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+static void testAzureIoTHubClient_Connect_InvalidArgFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    bool xSessionPresent;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    assert_int_equal( AzureIoTHubClient_Connect( NULL,
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:234]: comments are missing

File: tests/azure_iot_hub_client_ut.c [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+    bool xSessionPresent;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Connect, eAzureIoTMQTTServerRefused );
+    assert_int_not_equal( AzureIoTHubClient_Connect( &xTestIoTHubClient,
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:257]: equal

File: tests/azure_iot_hub_client_ut.c [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+static void testAzureIoTHubClient_Disconnect_MQTTDisconnectFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Disconnect, eAzureIoTMQTTIllegalState );
+    assert_int_not_equal( AzureIoTHubClient_Disconnect( &xTestIoTHubClient ),
```
Comment (hihigupt) [inline tests/azure_iot_hub_client_ut.c:298]: similar feedback for all not_equal

**Comment by hihigupt** [comment]
> @hihigupt the vast majority of the low code cov is the functions that aren't used yet.

(y)

**Review by jspaith — APPROVED** [review]

**Review by hihigupt — APPROVED** [review]

**Pull request merged**